hashing/src/error.rs

```diff
@@ -0,0 +1,32 @@
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+#[allow(unused)]
+pub enum MerkleVerificationError {
+    #[error("Index out of bounds. Count: {count}, index: {index}")]
+    IndexOutOfBounds { count: u64, index: u64 },
+    #[error(
+        "Unexpected proof length. Count: {count}, index: {index}, \
+         expected proof length: {expected_proof_length}, \
+         actual proof length: {actual_proof_length}"
+    )]
+    UnexpectedProofLength {
+        count: u64,
+        index: u64,
+        expected_proof_length: u8,
+        actual_proof_length: usize,
+    },
+}
+
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum MerkleConstructionError {
+    #[error("Could not construct Merkle proof. Empty proof must have index 0. Index: {index}")]
+    EmptyProofMustHaveAtLeastOneEntry { index: u64 },
```

Comment:
```suggestion
    #[error("Could not construct Merkle proof with no leaves")]
    ProofMustHaveAtLeastOneEntry { index: u64 },
```

Although… from outside the crate, this can't be triggered at all anymore, if we don't expose `InexedMerkleProof`. How about just returning `IndexOutOfBounds` in that case, too? If `count` is 0, the `index` _must_ be `>= count` anyway, so that would also be appropriate?

Reply:
Variant removed.